src/randgraphs.jl

```diff
@@ -148,37 +148,81 @@ function _try_creation(n::Int, k::Vector{Int}, rng::AbstractRNG)
+with `k` edges that are preferentially attached to existing nodes with high degree.
+"""
+function barabasi_albert(n::Integer, n0::Integer, k::Integer; is_directed::Bool = false, complete::Bool = false, keyargs...)
+    if complete
+        g = is_directed ? CompleteDiGraph(n0) : CompleteGraph(n0)
+    else
+        g = is_directed ? DiGraph(n0) : Graph(n0)
+    end
+    
+    barabasi_albert!(g, n, k; keyargs...)
+    return g
+end
 
-Creates a [Barabási–Albert model](https://en.wikipedia.org/wiki/Barab%C3%A1si%E2%80%93Albert_model)
-random graph with `n` nodes is grown by attaching new nodes each with `k` edges that
-are preferentially attached to existing nodes with high degree. Undirected graphs are
-created by default; use `is_directed=true` to override.
 """
-function barabasi_albert(n::Integer, k::Integer; is_directed=false, seed::Int=-1)
-    @assert(1<=k<n, "Barabási–Albert network must have 1 <= k < n")
-    g = is_directed ? DiGraph(n) : Graph(n)
+barabasi_albert!(g::Union{Graph,DiGraph}, n::Integer, k::Integer; seed::Int = -1)
+Creates a [Barabási–Albert model](https://en.wikipedia.org/wiki/Barab%C3%A1si%E2%80%93Albert_model) 
+random graph with `n` nodes that is grown by attaching new nodes to existing graph `g`, each
+with `k` edges that are preferentially attached to existing nodes with high degree.
+"""
+function barabasi_albert!(g::Union{Graph,DiGraph}, n::Integer, k::Integer; seed::Int=-1)
+    n0 = nv(g)
+    1 <= k <= n0 <= n || throw(ArgumentError("Barabási-Albert model requires 1 <= k <= n0 <= n where n0 is the number of nodes in graph g"))
+    n0 == n && return g
+    
+    # seed random number generator
     seed > 0 && srand(seed)
-    # Target nodes for new edges
-    targets = collect(1:k)
-    # List of existing nodes, with nodes repeated once for each adjacent edge
+
+    # add missing vertices
+    sizehint!(g.fadjlist, n)
+    add_vertices!(g, n - n0)
+
+    # vector of targets
+    targets = Vector{Int}(k)
+
+    # if graph doesn't contain edges add k edges from node n0+1
+    # this ensures that the sum of all weights is nonzero
+    if ne(g) == 0
+        # add links to targets
+        for target in sample!(collect(1:n0), k)
+            add_edge!(g, n0+1, target)
+        end
+
+        # prevents adding another k edges from node n0+1
+        n0 += 1
+    end
+
+    # list of existing nodes (each node is repeated once for each adjacent edge)
     repeated_nodes = Vector{Int}()
-    # Array to record the node is or not picked
-    node_status = fill(false, n)
-    sizehint!(repeated_nodes, (n-k)*k)
-    source = k + 1
-    while source <= n
-        for target in targets
-            # Add edges to k nodes from the source
-            add_edge!(g, source, target)
-            push!(repeated_nodes, source)
-            push!(repeated_nodes, target)
-            # Reset the node_status for target in targets
-            node_status[target] = false
+    sizehint!(repeated_nodes, 2*(n-n0)*k + 2*ne(g))
```

Comment:
What about `repeated_nodes = Array(Int, 2*(n-n0)*k + 2*ne(g))` and use `node_last` to record the number of existing nodes. Because we know the actual number of edges. Thus line 288 can be instead of `target = repeated_nodes[rand(1:node_last)]` or `target = repeated_nodes[StatsBase.randi(node_last)]`.